### ISSUE TITLE
Investigate faster algorithms for eccentricity-related concepts

### ISSUE THREAD

**Issue by szhorvat** [igraph/igraph#2684]
This paper proposes a faster algorithm for diameter/radius, and possibly generalizes more broadly to eccentricity-related concepts:

 * Michele Borassi, Pierluigi Crescenzi, Michel Habib, Walter A. Kosters, Andrea Marino, Frank W. Takes,
Fast diameter and radius BFS-based computation in (weakly connected) real-world graphs: With an application to the six degrees of separation games, Theoretical Computer Science, Volume 586, 2015, Pages 59-80 https://doi.org/10.1016/j.tcs.2015.02.033

NetworkX has an implementation, but is' been de-emphasized, and it's unclear why. See https://github.com/networkx/networkx/issues/7687

We should investigate if this is worth implementing in igraph.

Related: #2683 

@franktakes, let me know if you have any input on the topic. Are there any limitations we should be aware of? Is this method always better than the naive one, and can it serve as the default implementation?

**Comment by franktakes** [comment]
Correct, my algorithm almost always requires fewer BFSes than APSP, and in real-world networks (sparse, with a skewed degree dsitribution, small-world) this reduces the work from _n_ BFSes to a a handful or a few dozen BFSes required to find the diameter. This means orders of magnitude improvement in computation time if you think about large networks. 
To achieve this speedup in computing diamter and radius, there is some extra _O(n)_ bookkeeping going on after each BFS. That time is almost always neglectable compared to the BFSes themselves, which run in _O(m)_ each.

There are some worst-case examples, such as regular graphs (all nodes have the same degree, such as circle graphs) and (near)-complete graphs, where there is little to no speedup compared to the naive implementation. I would imagine that the bookkeeping then makes it a tiny bit slower than a naive implementation. 

Indeed, I once had the algorithm incorporated into networkx, and there it's only used if you use `useBounds = True`. Thanks for pointing me at the thread where they now discuss this, I'll keep an eye on this one and that one as well!

**Comment by GrishaVar** [comment]
@szhorvat I'll start work on this

**Comment by szhorvat** [comment]
Thanks @GrishaVar !

Note for observers: @GrishaVar is a student in my network science class and we are communicating about this project offline as well.
